### PR TITLE
feat: New Redis source and sink

### PR DESCRIPTION
**Pull request by shumin1027**
I creat a plugin to support redis(list、channel) as a data source and sink 

Context Issue: #3319 

the example config:
```toml
[sources.in]
type = "redis"
url = "redis://127.0.0.1:6379/0"
data_type = "list"
key = "vector"
method = "brpop"
redis_key = "redis_key"


[sinks.out]
type = "redis"
url = "redis://127.0.0.1:6379/0"
data_type = "list"
key = "vector"
method = "lpush"
healthcheck = true
inputs = ["in"]
encoding.codec = "json"
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,6 @@ scripts/package-lock.json
 target
 node_modules
 tests/data/wasm/*/target
+
+
+Cargo.lock
```
Comment (jamtur01) [inline .gitignore:18]: Was this change intended? Either way, the file needs to end in a new line.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+stages:
```
Comment (jamtur01) [inline .gitlab-ci.yml:1]: Also, this file seems inadvertently added.

File: Cargo.toml [hunk 3]
```diff
@@ -531,7 +538,7 @@ prometheus-integration-tests = ["sinks-prometheus", "sources-prometheus", "bytes
 pulsar-integration-tests = ["sinks-pulsar"]
 splunk-integration-tests = ["sinks-splunk_hec", "warp"]
 
-shutdown-tests = ["sources","sinks-console","sinks-prometheus","sinks-blackhole","unix","rdkafka","transforms-log_to_metric","transforms-lua"]
+shutdown-tests = ["sources", "sinks-console", "sinks-prometheus", "sinks-blackhole", "unix", "rdkafka", "transforms-log_to_metric", "transforms-lua"]
```
Comment (jamtur01) [inline Cargo.toml:541]: The spaces between the items also seem to be added inadvertently.

File: src/sinks/redis.rs [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+            Type::List => {
+                let _ = std::mem::replace(
+                    &mut self.state,
+                    RedisSinkState::Sending(Box::pin(async move {
+                        let result = lpush((*conn).clone(), key.clone(), encoded.clone()).await;
+                        (conn, result)
+                    })),
+                );
+
+                // self.in_flight.push(Box::pin(async move {
```
Comment (jamtur01) [inline src/sinks/redis.rs:198]: Any reason for the large blocks of code commented out?

File: src/sources/redis/list.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    client: redis::Client,
+    key: String,
+    mut shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> Source {
+    let mut out = out.sink_map_err(|e| error!("error sending event: {:?}", e));
+
+    let fut = async move {
+        info!("get redis connection manager ……");
+        // let mut conn = client.get_async_connection().await.expect("failed to get redis async connection!");
```
Comment (jamtur01) [inline src/sources/redis/list.rs:30]: line commented out?

File: src/sources/redis/list.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+fn create_event(
+    line: &str,
+    _key: &str,
+) -> Event {
+    let mut event = Event::from(line);
+    // Add source type
+    event.as_mut_log().insert(log_schema().source_type_key(), Bytes::from("redis"));
+    // event.as_mut_log().insert("redis_key", key);
```
Comment (jamtur01) [inline src/sources/redis/list.rs:64]: Same?

File: src/sources/redis/mod.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            Ok(list::watch(client, key, shutdown, out))
+        }
+    }
+}
+
+fn build_client(config: &RedisSourceConfig) -> crate::Result<redis::Client> {
+    info!("open redis client ……");
+    let client = redis::Client::open(config.url.as_str())?;
+    info!("open redis client successed!");
+    // let mut conn = client.get_connection()?;
```
Comment (jamtur01) [inline src/sources/redis/mod.rs:101]: Same?

**Comment by jamtur01** [comment]
@shumin1027 Thanks for your contribution! I put some high-level comments in about the PR overall.

**Comment by jamtur01** [comment]
@shumin1027 Hi! Thanks for the updates. Can you please also make sure your commits are signed so our DCO is satisfied? Thanks!

**Comment by jamtur01** [comment]
There are some failed formatting checks - you can run all of those locally before you push if that helps.

**Comment by shumin1027** [comment]
@jamtur01 Thanks for your suggestions，I will remove the unused commented lines and fmt code

**Review by fanatid — CHANGES_REQUESTED** [review]
While it's ok for initial implementation do not have internal events I think it will better to add them.
Need tests / integration tests and docs.

File: src/sinks/redis.rs [hunk 8]
```diff
@@ -0,0 +1,282 @@
+extern crate redis;
+
+use crate::{
+    buffers::Acker,
+    config::{log_schema, DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
+    event::Event,
+    sinks::util::encoding::{EncodingConfig, EncodingConfigWithDefault, EncodingConfiguration},
+};
+use futures::{future::BoxFuture, ready, stream::FuturesUnordered, FutureExt, Sink, Stream};
+use redis::{aio::ConnectionManager, AsyncCommands, RedisError, RedisFuture, RedisResult};
```
[the remark this answers is not in the document]
Comment (shumin1027) [reply src/sinks/redis.rs:10]: ConnectionManager implement the [ConnectionLike](https://docs.rs/redis/0.17.0/redis/aio/trait.ConnectionLike.html) trait and the Send & Sized trait ,so the `ConnectionManager` can be used as `AsyncCommands`
[AsyncCommands](https://docs.rs/redis/0.17.0/redis/trait.AsyncCommands.html)
[AsyncCommands Src](https://docs.rs/redis/0.17.0/src/redis/commands.rs.html#1975)
```rust
#[cfg(feature = "aio")]
impl<T> AsyncCommands for T where T: crate::aio::ConnectionLike + Send + Sized {}

```

File: src/sinks/redis.rs [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            Type::Channel => {
+                let _ = std::mem::replace(
+                    &mut self.state,
+                    RedisSinkState::Sending(Box::pin(async move {
+                        let result = publish((*conn).clone(), key.clone(), encoded.clone()).await;
+                        (conn, result)
+                    })),
+                );
+            }
+        }
```
[the remark this answers is not in the document]
Comment (shumin1027) [reply src/sinks/redis.rs:195]: At first I planned also support `redis stream`， considering that  `redis stream` and `kafka` are very similar，
and I haven't particularly understood how to write a vector plugin，so I refer to the implementation of `kafka` and `pulsar`

**Comment by shumin1027** [comment]
@fanatid Thanks，I will add some tests / integration tests and docs

**Comment by shumin1027** [comment]
@fanatid I just started learning rust， not yet particularly familiar with the whole project and how to do integration testing，
I am currently maintaining this function(redis source and sink) myself, so I need to constantly update the code to adapt to the  code(vector) update，can you help me to let this function be better maintained ?

**Comment by fanatid** [comment]
This code looks very good for a start with rust :slightly_smiling_face: we would happy to have redis source/sink in vector, do not hesitate to ask any rust / other questions on our discord server.
integration tests: it looks very similar to the usual tests but will need real redis for communication, I think a good example will be PR with nginx (https://github.com/timberio/vector/pull/5209/files `ctrl+f: integration` :smile:)
What do you think about split this PR into two: source and sink? I think we would able to finish components faster one by one than both in one place.

**Comment by jamtur01** [comment]
@shumin1027 You don't need to add yourself to CODEOWNERS - we generally handle that internally to the project. Thanks.

**Comment by shumin1027** [comment]
@fanatid  thanks for your guidance，I have add redis sink&source integration test

**Review by jszwedko — COMMENTED** [review]
This is great, thanks so much for the contribution @shumin1027.

I left some in-line comments below, mostly around logging and better handling of error conditions.

We should also add internal events for the sink as you did with the source.

I agree with @fanatid that splitting this into two PRs, one for the source and one for the sink may allow them to get merged more quickly. It's up to you though.

We will want to add cue-based docs (see `docs/` folder), but we can help with that when this is ready.

File: src/sinks/redis.rs [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+inventory::submit! {
+    SinkDescription::new::<RedisSinkConfig>("redis")
+}
+
+impl GenerateConfig for RedisSinkConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(&Self {
+            url: "redis://localhost/0".to_owned(),
+            key: "vector".to_owned(),
+            encoding: Encoding::Json.into(),
```
[the remark this answers is not in the document]
Comment (fanatid) [reply src/sinks/redis.rs:71]: +1 for `Encoding::Json` by default or remove default value at all. `Text` by default will be removed from sinks soon https://github.com/timberio/vector/pull/5281

File: src/sinks/redis.rs [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+            match ready!(Pin::new(&mut this.in_flight).poll_next(cx)) {
+                Some((seqno, Ok(result))) => {
+                    trace!(
+                        message = "Redis sink produced message.",
+                        length = ?result,
+                    );
+
+                    this.pending_acks.insert(seqno);
+                    let mut num_to_ack = 0;
+                    while this.pending_acks.remove(&this.seq_tail) {
```
[the remark this answers is not in the document]
Comment (fanatid) [reply src/sinks/redis.rs:205]: Example with A, B, C, D, E looks correct for me. What numbers of events for acking you would expect to see @jszwedko?

[the remark this answers is not in the document]
Comment (jszwedko) [reply src/sinks/redis.rs:205]: Sorry, I mean that they would ack in the wrong order. As far as I can tell, the acker just takes a number of messages to ack. I'm assuming (but need to verify) that it treats them in FIFO order. If Redis processes them in a different order we could end up acking a message in vector that hasn't actually been processed.

Imagine, in my example, that `vector` died after processing event C. If we had ack'd 1 every time, then B would have been ack'd even though it maybe was never successfully processed by redis. I think, in that case, we'd want to ideally keep it in the buffer and resend it when vector comes back up, even if that means re-sending event D too.

[the remark this answers is not in the document]
Comment (fanatid) [reply src/sinks/redis.rs:205]: @lukesteensen we can make `pending_acks` and acking more clear if change `FuturesUnordered` to `FuturesOrdered`, but I'm not sure will change slowing acking or not. Should we try?

[the remark this answers is not in the document]
Comment (jszwedko) [reply src/sinks/redis.rs:205]: :+1: I talked with @fanatid out-of-band about this. I didn't realize this was a pattern we were already using and I can see how `seq_head`/`seq_tail` keeps us acking in the right order.

I also agree that using `FuturesOrdered` would make this more clear. It seems to do what we want (run futures in parallel, but only resolve them in order).

**Comment by shumin1027** [comment]
@jszwedko @fanatid 
Thanks for your suggestions，let me know how to improve the code，I will try to split this into two PRs, one for the source and one for the sink

**Comment by binarylogic** [comment]
@shumin1027 should we close this in favor of 2 separate PRs?

**Comment by shumin1027** [comment]
@binarylogic Ok, you can close this and I will push PRs for this two issues: 


- #5696 add redis source support 
- #5697 add redis sink support 

Thank you！

**Comment by leebenson** [comment]
@shumin1027, just a quick update -- I've assigned myself as the reviewer for #5696 and #5697.

I don't currently have a firm ETA for review, but this is now on my radar and I'll keep you posted.

Many thanks for your contribution!

I'll go ahead and close this PR, given the separate source/sink PRs above. Thanks!

**Comment by shumin1027** [comment]
@leebenson  
thank you ，I create a new  PR: #7078

**Pull request closed without merging**